plugin.py

```diff
@@ -46,77 +45,83 @@ def prepare(self):
         self.module_input = self.board_prefix + '.dsn'
         self.module_output = self.board_prefix + '.ses'
         
+        self.module_command = [self.java_path, "-jar", self.module_path, "-de", self.module_input, "-s"]
+        
         if os.path.isfile(self.module_input):
             os.remove(self.module_input)
         if os.path.isfile(self.module_output):
             os.remove(self.module_output)
-
+        
     # export board.dsn file from pcbnew
     def RunExport(self):
         ok = pcbnew.ExportSpecctraDSN(self.module_input)
         if ok and os.path.isfile(self.module_input):
             return True
         else:
-            show_error("""
+            wx_show_error("""
             Failed to invoke:
             * pcbnew.ExportSpecctraDSN
             """)
             return False
-        
+
     # auto route by invoking FreeRouting.jar
     def RunRouter(self):
-        command = [self.java_path, "-jar", self.module_path, "-de", self.module_input, "-s"]
-        try:
-            process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-            output, error = process.communicate()
-            if process.returncode == 0:
+        command = self.module_command
+        dialog = wx_make_cancel("""
+        Complete or Terminate FreeRouting:
+        * to complete, close Java window
+        * to terminate, press Terminate here
+        """)
+        invoker = ProcessThread(command, dialog)
+        dialog.Show()  # dialog first
+        invoker.start()  # run java process
+        result = dialog.ShowModal()  # block pcbnew here
+        if result == wx.ID_OK:  # return via terminate button
+            invoker.terminate()
+            return False
+        elif result == wx.ID_CANCEL:  # return via dialog.Destroy()
+            if invoker.has_ok():
                 return True
-            else:  
-                show_error("""
-                Java program failure:
-                exit code: %d
-                --- std output ---
-                %s
-                --- std error  ---
-                %s
-                """ % (process.returncode, output, error))
-                return False
-        except Exception as error:
-                show_error("""
-                Java process failure: 
-                %s""" % str(error))
+            else:
+                invoker.show_error()
                 return False
+        else:
+            return False  # should not happen
```

Comment:
It's pretty idiomatic to write
```
if result == wx.ID_OK:
  ...
else:
  ...
```
Also it's best to destroy dialog here where you create it. No need to do it in CallAfter().

Reply:
> No need to do it in CallAfter()

tried that, it crashes kicad

Reply:
On windows this does not work. Dialog never closes even after java process completes.

What you would do normally do is call dialog.EndModal(resultcode) where you want to close it and destroy it right where you call dialog.ShowModal() to not leak it. 
Problem is wxMessageDialog is not actually a wxDialog so it ignores EndModal. 
To do this properly you will have to make an instance of wx.Dialog yourself.